src/compiler/checker.ts

```diff
@@ -18176,17 +18404,44 @@ namespace ts {
             return !!(getObjectFlags(type) & ObjectFlags.Reference && (<TypeReference>type).target.objectFlags & ObjectFlags.Tuple);
         }
 
+        function isGenericTupleType(type: Type): type is TupleTypeReference {
+            return isTupleType(type) && !!(type.target.combinedFlags & ElementFlags.Variadic);
+        }
+
+        function isSingleElementGenericTupleType(type: Type): type is TupleTypeReference {
+            return isGenericTupleType(type) && type.target.elementFlags.length === 1;
+        }
+
         function getRestTypeOfTupleType(type: TupleTypeReference) {
-            return type.target.hasRestElement ? getTypeArguments(type)[type.target.typeParameters!.length - 1] : undefined;
+            return getElementTypeOfTupleType(type, type.target.fixedLength);
         }
 
         function getRestArrayTypeOfTupleType(type: TupleTypeReference) {
             const restType = getRestTypeOfTupleType(type);
             return restType && createArrayType(restType);
         }
 
-        function getLengthOfTupleType(type: TupleTypeReference) {
-            return getTypeReferenceArity(type) - (type.target.hasRestElement ? 1 : 0);
+        function getEndLengthOfType(type: Type) {
+            return isTupleType(type) ? getTypeReferenceArity(type) - findLastIndex(type.target.elementFlags, f => !!(f & ElementFlags.Variable)) - 1 : 0;
+        }
+
+        function getElementTypeOfTupleType(type: TupleTypeReference, index: number, endSkipCount = 0, writing = false) {

```

Comment:
```suggestion
        function getElementTypeOfSliceOfTupleType(type: TupleTypeReference, start: number, end: number, writing = false) {

```
might be an easier to grok signature here. What this does (get the combiend element type for a slice of a tuple) isn't obvious from the current name, and the `endSkipCount` parameter feels somewhat awkward, when framed in the context of slicing (why not provide the last element index instead?).

Reply:
Sure, I can rename it. The `endSkipCount` simply ended up being more convenient as we're generally reasoning about elements towards the middle from the start or from the end.